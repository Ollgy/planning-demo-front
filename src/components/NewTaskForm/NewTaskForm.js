import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import CloseButton from '../CloseButton';
import Button from '../Button';
import Input from '../Input';
import Textarea from '../Textarea';
import FormMsg from '../FormMsg';
import getformattedTime from './utils/formattedTime';
import {
  inputForm,
  setFields,
  setInitialState,
  setMsgForm,
  validateForm,
  clearError,
  getUser,
  saveNewTask
} from '../../modules/NewTaskForm';
import { F, M } from '../../const';
import styles from './NewTaskForm.module.css';
import texts from '../../texts';
import formStyles from '../LoginForm/LoginForm.module.css';

class NewTaskForm extends PureComponent {
  constructor(props) {
    super(props);
    this.activeInput = React.createRef();
  }

  componentWillUnmount() {
    this.props.setInitialState();
  }

  async componentDidMount() {
    let { taskData, mode, getUser, setFields } = this.props;

    if (mode === M.create) {
      await getUser(taskData.executorId);
    }

    setFields({
      task: taskData.task || '',
      comment: taskData.comment || ''
    });

    this.activeInput.current.focus();
  }

  onInputChange = e => {
    const { inputForm, clearError } = this.props;
    const { value, name } = e.target;

    inputForm({ input: name, value });
    clearError({ input: name });
  };

  onSubmit = async e => {
    e.preventDefault();

    const mode = e.target.name;

    const {
      fields,
      taskData,
      validateForm,
      saveNewTask,
      updateTask,
      setMsgForm
    } = this.props;

    await validateForm(fields);

    const { validation } = this.props;

    if (validation.status && mode === M.create) {
      await saveNewTask({
        ...fields,
        date: {
          create: getformattedTime(),
          begin: '',
          done: ''
        }
      });
      setMsgForm(texts.addTask);
    } else if (validation.status && mode === M.update) {
      await updateTask(taskData.id, fields);
      setMsgForm(texts.updTask);
    }

    if (this.activeInput) {
      this.activeInput.current.focus();
    }
  };

  render() {
    const { fields, validation, mode, onClose } = this.props;
    const { errorInputs, error } = validation;

    return (
      <div className={styles.container}>
        <div className={styles.formWrap}>
          <form action="/" method="post" onSubmit={this.onSubmit} name={mode}>
            <span
              className={styles.title}
              children={`Сотрудник: ${fields.executorName}`}
            />
            <div
              className={`${formStyles.formItem} ${formStyles.formItemWide}`}
              key={F.task}
            >
              <Input
                type="text"
                ref={this.activeInput}
                name={F.task}
                value={fields[F.task]}
                readonly={false}
                placeholder=". . ."
                errorClass={errorInputs.includes(F.task) ? error : null}
                onChange={this.onInputChange}
              />
              <label className={formStyles.formLabel} htmlFor={F.task}>
                <span>Задача</span>
              </label>
            </div>
            <div
              className={`${formStyles.formItem} ${formStyles.formItemColumn}`}
              key={F.comment}
            >
              <Textarea
                size="big"
                name={F.comment}
                value={fields[F.comment]}
                readonly={false}
                placeholder=". . ."
                onChange={this.onInputChange}
              />
              <label className={formStyles.formLabel} htmlFor={F.comment}>
                <span>{texts.commentTask}</span>
              </label>
            </div>
            <div className={styles.msgBlock}>
              <FormMsg children={validation.msg} />
            </div>
            <div className={styles.buttonBlock}>
              {mode === M.create ? (
                <Button
                  type="submit"
                  name="create"
                  children="Добавить задачу"
                />
              ) : (
                <Button type="submit" name="update" children="Редактировать" />
              )}
            </div>
          </form>
          <div className={styles.closeButtonBlock}>
            <CloseButton size="s" onClick={onClose} />
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  fields: state.newTaskForm.datalist.fields,
  validation: state.newTaskForm.datalist.validation,
  executor: state.newTaskForm.executor.user
});

const mapDispatchToProps = dispatch => ({
  saveNewTask: arg => dispatch(saveNewTask(arg)),
  getUser: arg => dispatch(getUser(arg)),
  setFields: arg => dispatch(setFields(arg)),
  setInitialState: () => dispatch(setInitialState()),
  setMsgForm: arg => dispatch(setMsgForm(arg)),
  inputForm: arg => dispatch(inputForm(arg)),
  validateForm: arg => dispatch(validateForm(arg)),
  clearError: arg => dispatch(clearError(arg))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(NewTaskForm);
