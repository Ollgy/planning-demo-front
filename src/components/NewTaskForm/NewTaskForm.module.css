.container {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  background: rgba(255, 255, 255, 0.8);
  z-index: 10;
}

.formWrap {
  position: relative;
  width: 365px;
  padding: 35px;
  border: 2px solid #bfbfbf;
  border-radius: 15px;
  box-shadow: 5px 10px 10px 2px #e6e6e6, -5px 0px 10px 2px #e6e6e6;
  background-color: #fff;
  overflow: hidden;
}

.title {
  display: inline-block;
  color: #757575;
  margin-bottom: 30px;
}

.msgBlock {
  height: 40px;
  margin-bottom: 15px;
}

.buttonBlock {
}

.closeButtonBlock {
  position: absolute;
  top: 5px;
  right: 5px;
}
