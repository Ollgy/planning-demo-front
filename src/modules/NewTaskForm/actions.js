import { createAction } from 'redux-actions';
import api from '../../api';

export const validateForm = createAction('VALIDATE_FORM_TASK');
export const inputForm = createAction('INPUT_FORM_TASK');
export const clearError = createAction('CLEAR_ERROR_TASK');
export const setFields = createAction('SET_FIELDS_TASK');
export const setMsgForm = createAction('SET_MSG_FORM_TASK');
export const setInitialState = createAction('SET_INITIAL_STATE_TASK');

export const saveNewTaskRequest = createAction('SAVE_NEW_TASK_FORM_REQUEST');
export const saveNewTaskSuccess = createAction('SAVE_NEW_TASK_FORM_SUCCESS');
export const saveNewTaskFailure = createAction('SAVE_NEW_TASK_FORM_FAILURE');

export const saveNewTask = (userId, taskData) => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(saveNewTaskRequest());

      api
        .saveNewTask(userId, taskData)
        .then(task => {
          dispatch(saveNewTaskSuccess(task));
          resolve();
        })
        .catch(error => {
          dispatch(saveNewTaskFailure(error));
          reject();
        });
    });
};

export const getUserRequest = createAction('GET_USER_REQUEST');
export const getUserSuccess = createAction('GET_USER_SUCCESS');
export const getUserFailure = createAction('GET_USER_FAILURE');

export const getUser = userId => {
  return (dispatch, getState) =>
    new Promise((resolve, reject) => {
      dispatch(getUserRequest());

      api
        .getUser(userId)
        .then(user => {
          dispatch(getUserSuccess(user));
          resolve();
        })
        .catch(error => {
          dispatch(getUserFailure(error));
          reject();
        });
    });
};
