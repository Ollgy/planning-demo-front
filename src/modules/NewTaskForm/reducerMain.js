import { combineReducers } from 'redux';
import executor from './reducers/executor';
import newTask from './reducers/newTask';
import datalist from './reducers/datalist';

export default combineReducers({
  datalist,
  executor,
  newTask
});
