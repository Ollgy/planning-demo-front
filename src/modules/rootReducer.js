import { combineReducers } from 'redux';
import authFromToken from './Auth';
import login from './Login';
import loginForm from './LoginForm';
import profile from './Profile';
import personalForm from './PersonalForm';
import changePswrdForm from './ChangePswrdForm';
import newTaskForm from './NewTaskForm';
import notes from './Notes';
import dashboard from './Dashboard';
import staff from './Staff';
import cleanBase from './CleanBase';

export default combineReducers({
  authFromToken,
  login,
  loginForm,
  profile,
  personalForm,
  changePswrdForm,
  newTaskForm,
  notes,
  dashboard,
  staff,
  cleanBase
});
