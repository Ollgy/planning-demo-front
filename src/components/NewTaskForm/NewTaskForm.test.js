import React from 'react';
import NewTaskForm from './NewTaskForm';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import { createStore, compose, applyMiddleware } from 'redux';
import rootReducer from '../../modules/rootReducer';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, {}, compose(applyMiddleware(thunk)));

describe('<NewTaskForm/> UI component', () => {
  const shallowExpect = compose(
    expect,
    toJSON,
    shallow
  );

  it('Renders correctly', () => {
    shallowExpect(<NewTaskForm store={store} />).toMatchSnapshot();
  });

  const wrapper = shallow(<NewTaskForm store={store} />);
  const inner = wrapper.children().first();

  it('contains fields in props', () => {
    expect(inner.props().fields).toBeTruthy();
  });

  it('contains validation in props', () => {
    expect(inner.props().validation).toBeTruthy();
  });
});
