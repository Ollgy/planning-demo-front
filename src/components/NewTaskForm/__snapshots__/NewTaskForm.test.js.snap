// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewTaskForm/> UI component Renders correctly 1`] = `
<ContextProvider
  value={null}
>
  <NewTaskForm
    clearError={[Function]}
    executor={null}
    fields={
      Object {
        "comment": "",
        "executorId": "",
        "executorName": "",
        "task": "",
      }
    }
    getUser={[Function]}
    inputForm={[Function]}
    saveNewTask={[Function]}
    setFields={[Function]}
    setInitialState={[Function]}
    setMsgForm={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    validateForm={[Function]}
    validation={
      Object {
        "error": "",
        "errorInputs": Array [],
        "msg": "",
        "status": false,
      }
    }
  />
</ContextProvider>
`;
